src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs

```diff
@@ -0,0 +1,461 @@
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns>A task that completes with the result of <paramref name="operation"/>, after the semaphore has been exited.</returns>
+        public abstract Task ExecuteAsync(Func<Task> operation, CancellationToken cancellationToken = default);
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Disposes managed and unmanaged resources held by this instance.
+        /// </summary>
+        /// <param name="disposing"><c>true</c> if <see cref="Dispose()"/> was called; <c>false</c> if the object is being finalized.</param>
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                this.semaphore.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Throws an exception if this instance has been faulted.
+        /// </summary>
+        protected virtual void ThrowIfFaulted()
+        {
+        }
+
+        /// <summary>
+        /// Disposes the specfied release, swallowing certain exceptions.
+        /// </summary>
+        /// <param name="releaser">The releaser to dispose.</param>
+        private static void DisposeReleaserNoThrow(AsyncSemaphore.Releaser releaser)
+        {
+            try
+            {
+                releaser.Dispose();
+            }
+            catch (ObjectDisposedException)
+            {
+                // Swallow this, since in releasing the semaphore if it's already disposed the caller probably doesn't care.
+            }
+        }
+
+        /// <summary>
+        /// An implementation of <see cref="ReentrantSemaphore"/> supporting the <see cref="ReentrancyMode.NotRecognized"/> mode.
+        /// </summary>
+        private class NotRecognizedSemaphore : ReentrantSemaphore
+        {
+            /// <summary>
+            /// Initializes a new instance of the <see cref="NotRecognizedSemaphore"/> class.
+            /// </summary>
+            /// <param name="initialCount">The initial number of concurrent operations to allow.</param>
+            /// <param name="joinableTaskContext">The <see cref="JoinableTaskContext"/> to use to mitigate deadlocks.</param>
+            internal NotRecognizedSemaphore(int initialCount, JoinableTaskContext joinableTaskContext)
+                : base(initialCount, joinableTaskContext)
+            {
+            }
+
+            /// <inheritdoc />
+            public override async Task ExecuteAsync(Func<Task> operation, CancellationToken cancellationToken = default)
+            {
+                Requires.NotNull(operation, nameof(operation));
+                this.ThrowIfFaulted();
+
+                Func<Task> semaphoreUser = async delegate
+                {
+                    using (this.joinableTaskCollection?.Join())
```

Comment:
>joinableTaskCollection [](start = 32, length = 22)

Under what conditions would ```this.joinableTaskCollection``` or ```this.joinableTaskFactory``` (handled specially below) be null?

Reply:
Some apps don't need to use JTF (e.g. web applications or console apps), so they're allowed to pass null in.